Extract more-complex chapters from the pinned comment; select "normal" chapters or comment chapters

Lazyt3ch opened the pull request:
Hi Ilya,

I've found a case where each comment line can contain either one or more chapters, or no chapters at all: https://www.youtube.com/watch?v=PzlLXQ3RaDs
I've added some code to extract chapters from such comments.

I've also found a case where the chapters provided in the pinned comment are much better than the "key moments" automatically added by YouTube: https://www.youtube.com/watch?v=5Gt61EX6HZ4
In that case, when the extension tries to fetch chapters, it gets the "key moments" instead – and displays them as chapters. Unfortunately, the result makes no sense.
I've made minor changes to the extension, and now, if both the "normal" chapters (that is, either full-fledged chapters or "key moments") and the chapters extracted from the pinned comment are available, it select the more-numerous ones.

Actually, both cases are pretty unusual, so I'm not sure if handling them is worth the effort.

Regards,
Rashid

ris58h commented:
I seems like very specific case. I prefer to not complicate the codebase to solve rare/specific cases.

I believe, we should fix `getTimestampContexts` method. There are some flaws in the logic:
- `"\r\n"` instead of simple `"\n"`
- `return` when no all lines where processed

Also we could refactor `fetchChaptersFromComments` and other places where `fori` loop isn't necessary and could be replaced with map/filter.

Lazyt3ch commented:
> It seems like very specific case. I prefer to not complicate the codebase to solve rare/specific cases.

As you wish. The cases I described are very rare indeed.

> 
> I believe, we should fix `getTimestampContexts` method. There are some flaws in the logic:
> 
> * `"\r\n"` instead of simple `"\n"`
> * `return` when not all lines where processed. Shouldn't there be `continue` instead?
> 

Sure, I can do it. I've already replaced the "\r\n" string with the /\r?\n/ regex in the modified version.

> Also we could refactor `fetchChaptersFromComments` and other places where `fori` loop isn't necessary and could be replaced with map/filter.

I've already replaced the for-loop with the Array.prototype.find method in the fetchChaptersFromComments function in the modified version. I can do the same in the original version.
